django_orghierarchy/models.py

```diff
@@ -9,7 +9,7 @@ class AbstractDataSource(models.Model):
     Abstract data source model that provides required fields
     for custom data source model.
     """
-    name = models.CharField(max_length=255)
+    name = models.CharField(max_length=255, unique=True)
```

Comment:
Hm, comparing to the Linkedevents DataSource model https://github.com/City-of-Helsinki/linkedevents/blob/master/events/models.py#L62 , I noticed that there the name actually might *not* be unique, because it can be a more descriptive string :(

Therefore, I guess we'll have to make the abstract model compatible with linkedevents in that the string `id` itself is the unique string, but the name is not. That's how Linkedevents object ids are generated. Generating ids from the non-unique Linkedevents `name` field will lead to problems, they don't necessarily have the same string as `id` in Linkedevents database :(

Reply:
For linkedevents, we don't need to use this abstract data source model, instead we can use the existing data source model in linkedevents directly.

But there's one problem, since we only have the data source name from the paatos REST API (https://api.hel.fi/paatos/v1/organization/), what should be the id for the data source since the id is not auto-created?

Reply:
Well, the problem here is that the abstract model assumes the name is unique, and we will generate ids for organizations by the name field, even if this assumption is no longer valid. Therefore, we cannot use this field as a source of organization id.

Reply:
Actually the string that the paatos API provides (`helsinki`) is the unique identifier of the data source (not the human-readable name) in the paatos system too: https://github.com/City-of-Helsinki/paatos/blob/master/decisions/models/base.py#L18 So we can use it as the `id` here, if it is not renamed in calling the management command.

Reply:
I see, so if I understand you correctly, we need to do following changes:

- make data source id field a Charfield, the name field should not be unique
- the value for organization id should be `<data_source_id>:<origin_id>`

Reply:
So yeah, this works for now, but plugging in to Linkedevents will fail because Organization ids will be generated from the wrong field in Linkedevents :(

Reply:
~- make data source id field a Charfield, the name field should not be unique~
:point_up_2:  I think we can add an extra identifier field as like in https://github.com/City-of-Helsinki/paatos/blob/master/decisions/models/base.py#L18

Reply:
Yeah, I thought about adding an extra identifier field too.

Then, however, plugging this in to Linkedevents will mean we will have to have a method in the plugged DataSource model that tells our app that in this case, we should use the `id` instead of the `identifier` field 😂

Reply:
Essentially, our `AbstractDataSource` determines the API the pluggable Datasource model must implement, and currently Linkedevents fails at that.

Reply:
We can, of course, look at other apps and add `identifier` to linkedevents too, if it looks like that is the more sensible and common way to add a unique identifier, and you think a primary key is not a good place for a unique identifier.

Reply:
We can account for both `identifier` and `id` field in the app, and don't need to change anything in linkedevents

Reply:
> We can, of course, look at other apps and add identifier to linkedevents too, if it looks like that is the more sensible and common way to add a unique identifier, and you think a primary key is not a good place for a unique identifier

Yes, but probably just my personal preference, generally I would try to avoid changing the default primary key and use it for other purposes as django has its own use of the field internally, for example, sometimes we need to add this kind of logic: https://github.com/City-of-Helsinki/django-orghierarchy/pull/7/files#diff-00d90d11e737fa569968afdc01d6058fR70. While using a separate identifier field does not have this problem.